File: LeetCode/1839. Longest Substring Of All Vowels in Order/solution.py
class Solution:
    def longestBeautifulSubstring(self, word: str) -> int:
        mp = {ch: i+1 for i, ch in enumerate("aeiou")}
        s, cnt, res = 0, 0, 0
        for ch in word:
            ch = mp[ch]
            if s <= ch <= s+1:
                cnt += 1
                s = ch 
            elif ch == 1:
                cnt = 1
                s = ch
            else:
                cnt = 0
                s = 0
            if s == 5:
                res = max(res, cnt)
        return res

class Solution:
    def longestBeautifulSubstring(self, word: str) -> int:
        res, cur = 0, 0
        pre = {'e':'a', 'i':'e', 'o':'i', 'u':'o'}
        for i, ch in enumerate(word):
            if ch == 'a':
                if i and word[i-1] != ch:
                    cur = 1
                else:
                    cur += 1
            else:
                if i and (word[i-1] == ch or word[i-1] == pre[ch]) and cur:
                    cur += 1
                else:
                    cur = 0
                if ch == 'u':
                    res = max(res, cur)
        return res

File: LeetCode/1839. Longest Substring Of All Vowels in Order/solution.cpp
class Solution {
public:
    int longestBeautifulSubstring(string word) {
        const auto n = word.size();

        int cnt = 1;
        int len = 1;
        int max_len = 0;
        for (int i = 1; i != n; ++i) {
            if (word[i - 1] == word[i]) {
                ++len;
            } else if (word[i - 1] < word[i]) {
                ++len;
                ++cnt;
            } else {
                cnt = 1;
                len = 1;
            }
            
            if (cnt == 5) {
                max_len = max(max_len, len);
            }
        }
        return max_len;
    }
};



class Solution {
public:
    int longestBeautifulSubstring(string word) {
        std::vector<int> mp(128);
        string s = "aeiou";
        for (int i = 0; i < 5; i++){
            mp[s[i]] = i;
        }
        int dp = 5, pre, res = 0, sz;
        for (auto ch : word){
            int cur = mp[ch];
            if (dp < 5 && (cur == pre || cur == pre+1)){
                dp = cur;
                sz++;
                if (dp == 4)
                    res = max(res, sz);
            }
            else{
                dp = 5;
                if (cur == 0){
                    dp = 0;
                    sz = 1;
                }
            }
            pre = cur;
            // cout << cur << " " << dp << " " << sz << endl;

        }
        return res;
    }
};

